### PR TITLE
Dictionaries - add support for notifications

### PR DESCRIPTION
**Pull request by edualonso**
Added support for change listeners.

Change sets contain:

- deletions count
- array with inserted keys, if any
- array with modified keys, if any

### CODE REVIEW

File: realm/realm-library/src/main/cpp/observable_dictionary_wrapper.hpp
Comment (clementetb) [inline realm/realm-library/src/main/cpp/observable_dictionary_wrapper.hpp:2]:
```suggestion
 * Copyright 2021 Realm Inc.
```

File: realm/realm-library/src/main/java/io/realm/MapChangeSet.java
Comment (clementetb) [inline realm/realm-library/src/main/java/io/realm/MapChangeSet.java:48]: We might like to have the same function name as for Lists or Objects: `getChanges`? What do you think?

**Review by clementetb — APPROVED** [review]
Awesome to see change listeners in place 🚀

File: realm/realm-library/src/main/java/io/realm/MapChangeSet.java
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/MapChangeSet.java:48]: Fair enough

**Comment by edualonso** [comment]
Pending debug with someone from Core regarding https://github.com/realm/realm-core/issues/4487 as it seems I'm doing the right thing in the test. I have created a ticket to revisit tests that have been either ignored or limited due to bugs: RJAVA-917

**Comment by edualonso** [comment]
I will merge this as I cannot wait much longer until I get help from core debugging the changesets and getting a second review given that we all are incredibly busy.

**Pull request merged**
